[zhekaiii] iP

zhekaiii opened the pull request:
### Checkbot ✅
> "This bot is great!" – classmate

Checkbot is a text-based chatbot that helps you organize and maintain your tasks. This is great for:
* Procrastinators
* Students

The following features are available:
- [x] Adding tasks
- [x] Deleting tasks
- [x] Marking tasks as complete/incomplete
- [ ] _More features to come!_

We have the following **types** of tasks:
1. Todo
2. Deadline (requires end date)
3. Event (requires start and end date)

Download an early release [here](https://github.com/zhekaiii/ip/releases/tag/Level-9)!

----------------------

Here is a code snippet in `Checkbot.java`:
```java
public void run() {
    ui.showWelcome();

    Scanner scanner = new Scanner(System.in);
    while (true) {
        String input = scanner.nextLine();
        try {
            Command c = parser.parse(input);
            c.execute(todoList, storage, ui);
            if (c.isBye()) {
                break;
            }
        } catch (CheckbotException e) {
            ui.showError(e);
        }
    }

    scanner.close();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

zhekaiii committed:
renamed Duke to Checkbot

zhekaiii committed:
add echo function to Checkbot

zhekaiii committed:
add function to add new tasks and list out todo list to Checkbot

zhekaiii committed:
add mark and unmark task functionality

zhekaiii committed:
add different types of tasks

zhekaiii committed:
add automated text ui testing

zhekaiii committed:
add exception handling for invalid inputs and update text ui test

zhekaiii committed:
add delete functionality to Checkbot and update test cases

zhekaiii committed:
Add functionality to save tasks in a txt file

zhekaiii committed:
Add ability to parse and reformat date strings

zhekaiii committed:
refactor: extract out Storage related functions as separate class

zhekaiii committed:
refactor: extract out ui and parsing related functions as separate classes

zhekaiii committed:
refactor: divide classes into packages

zhekaiii committed:
docs: add javadocs to most public classes and methods

zhekaiii committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

zhekaiii committed:
Update class name in build.gradle

zhekaiii committed:
test: add unit tests for AddCommand and Storage

zhekaiii committed:
build: package app as a jar file

zhekaiii committed:
Merge branch 'master' into branch-A-JavaDoc

zhekaiii committed:
refactor: replace wildcard imports

zhekaiii committed:
Add functionality to find tasks

zhekaiii committed:
Merge branch 'branch-A-CodingStandard'

zhekaiii committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/checkbot/Ui.java
#	src/main/java/checkbot/task/TodoList.java

zhekaiii committed:
fix: add missing import in Parser

zhekaiii committed:
Fix runtest.sh pointing to wrong directory and test case

zhekaiii committed:
Add more header comments to methods

willowisp01 commented on src/main/java/checkbot/Parser.java:
```diff
@@ -0,0 +1,125 @@
+                    throw new MissingDeadlineException();
+                }
+            }
+            Pattern pattern = Pattern.compile("event (.*) /(from|to)(.*) /(from|to)(.*)");
```
I wasn't aware of the Pattern class myself, but it looks like a great way of saving the effort of doing it manually!

willowisp01 commented on src/main/java/checkbot/Ui.java:
```diff
@@ -0,0 +1,118 @@
+     */
+    public void showWelcome() {
+        print(
+                "Hello, I'm Checkbot, your personal assistant.\n" +
```
[Break before the operator](https://se-education.org/guides/conventions/java/intermediate.html#layout)  (i.e. put the + on the next line)

willowisp01 reviewed (commented):
I think the code is really well done! Honestly it was quite hard to find any coding standard violation at all. I like the way the code is organized into different packages and classes, it shows a clear separation of concerns and is neat.

zhekaiii committed:
Add files for Checkstyle

zhekaiii committed:
Add JavaFX GUI to the application

zhekaiii committed:
Add missing image to repo

zhekaiii committed:
Add missing save function to MarkCommand

zhekaiii committed:
Add overwrite for Equals method for Tasks

zhekaiii committed:
Refactor TodoList.java

Replaces TodoList array with ArrayList and uses Stream methods inside `toString`, `formatForFile`, and `find`.

Stream methods allow us to write code that is cleaner and easier to read.

zhekaiii committed:
Refactor Parser and Storage

Reorders package imports in Parser.java and use try with resource instead for Scanner in Storage.java.

zhekaiii committed:
Refactor Parser and Storage

Reorders package imports in Parser.java and use try with resource instead for Scanner in Storage.java.

zhekaiii committed:
Fix CheckStyle issues

zhekaiii committed:
Fix CheckStyle issues

zhekaiii committed:
Rename Command::execute to Command::executeCommand

zhekaiii committed:
Extract out methods to improve readability

zhekaiii committed:
Refactor Parser.java to make happy path more prominent

zhekaiii committed:
Add assertions to the application

There are 2 cases of usage of "else" to execute the final option. Assertions have been added to those branches to make it clearer and allow us to detect if there is a bug.

zhekaiii committed:
Merge pull request #1 from zhekaiii/branch-A-CodeQuality

Improve code quality of the project

zhekaiii committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/checkbot/Parser.java

zhekaiii committed:
Merge pull request #2 from zhekaiii/branch-A-Streams

Adopt usage of Java Stream into the project

zhekaiii committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/checkbot/Parser.java
#	src/main/java/checkbot/Storage.java

zhekaiii committed:
Merge pull request #3 from zhekaiii/branch-A-Assertions

Add assertions to the application

yadunut commented on src/main/java/checkbot/Storage.java:
```diff
@@ -0,0 +1,123 @@
+            task = new Deadline(name, byWhen);
+        } else {
+            assert type.equals((EVENT_CODE)) : "Task should start with an 'E'";
+            Matcher eventMatcher = Pattern.compile("(.*) \\| (.*) \\| (.*)").matcher(taskDetails);
```
Nice work with regex. It would be good to have a little comment for complex regexes to give a brief overview of what it does!

zhekaiii replied:
Noted, will add comments to explain the regexp:)

yadunut commented on src/main/java/checkbot/Parser.java:
```diff
@@ -0,0 +1,158 @@
+
+    private static AddCommand parseAddCommand(String input) throws
+            EmptyDescriptionException, MissingDeadlineException, MissingFromException, MissingToException {
+        String[] splitString = input.split("todo|deadline|event|/(by|from|to)");
```
Nice work, TIL with the string.split with regex(?) support

yadunut commented on src/main/java/checkbot/task/TodoList.java:
```diff
@@ -0,0 +1,123 @@
+     */
+    public TodoList find(String substr) {
+        return new TodoList(
+                new ArrayList<>(List.of(
+                        taskList
+                                .stream()
+                                .filter(task -> task.nameContains(substr))
+                                .toArray(Task[]::new)))
+        );
```
Instead of this, you could do `taskList.stream().filter(...).collect(Collectors.toList())`

zhekaiii replied:
Ahhhh thank you for the suggestion!

yadunut reviewed (commented):
Very nice work, some comments. Nice work on the comprehensive exceptions and documentation! Made it very easy to parse your work

zhekaiii committed:
Add GitHub actions for CI

zhekaiii committed:
Fix checkstyle toolVersion

zhekaiii committed:
Add some comments and refactors

zhekaiii committed:
Create TimeString class to handle parsing of dates

zhekaiii committed:
Handle and prevent duplicate tasks

zhekaiii committed:
Merge pull request #4 from zhekaiii/branch-C-DetectDuplicates

branch-C-DetectDuplicates

zhekaiii committed:
Add new unit tests for AddCommand class

zhekaiii committed:
Add new unit tests for TimeString class

zhekaiii committed:
Add user guide

zhekaiii committed:
Extract methods from Ui class to conform to SLAP

zhekaiii committed:
Add fallback message for list if task list is empty

zhekaiii committed:
Add help command

zhekaiii committed:
Fix some bugs regarding jar files
